api/src/pcapi/celery_tasks/tasks.py

```diff
@@ -41,8 +87,31 @@ def task(payload: dict) -> None:
                     parsed_payload = model.parse_obj(parsed_payload)
                 except pydantic_v1.ValidationError as exp:
                     logger.error("could not deserialize object", extra={"exception": exp})
+            else:
+                parsed_payload = payload
 
-            f(parsed_payload)
+            if max_per_time_window:
+                try:
+                    with rate_limit(name, time_window_size, max_per_time_window):
+                        f(parsed_payload)
+                except RateLimitedError as exp:
+                    logger.error(
+                        "task rate-limited",
+                        extra={
+                            "time_window_size": exp.time_window_size,
+                            "max_per_time_window": exp.max_per_time_window,
+                            "current": exp.current,
+                        },
+                    )
+                    calls_over_limit = exp.current - exp.max_per_time_window
+                    nb_time_windows_to_skip = 1 + (calls_over_limit // exp.max_per_time_window)
+                    time_to_wait_before_retry = exp.time_window_size * nb_time_windows_to_skip
+                    if time_to_wait_before_retry < MAX_RETRY_DURATION:
+                        self.retry(exc=exp, countdown=time_to_wait_before_retry)
+                    else:
+                        self.retry(exc=exp, countdown=MAX_RETRY_DURATION)
```

Comment:
Cela me semble être tout simplement un `min` :
```suggestion
                    self.retry(exc=exp, countdown=min(time_to_wait_before_retry, MAX_RETRY_DURATION))
```
Est-il possible d'expliquer la raison de `MAX_RETRY_DURATION` ? Si on empile beaucoup de tâches, ou si le _rate limit_ est très bas, pourquoi devrait-on obligatoirement les rejouer toutes 20 minutes plus tard ? Est-ce lié à la configuration de celery ?

Reply:
C'est complétement un `min` donc j'ai changé.

L'idée du MAX_RETRY_DURATION c'est d'éviter d'empiler la tâche pour dans trop longtemps. Tant que la tâche est en retry elle est dans la mémoire du worker et on risque de la perdre si le worker se fait kill trop brutalement.
Cf https://docs.celeryq.dev/en/latest/userguide/calling.html#eta-and-countdown.

Je rajoute un petit commentaire au dessus de ces valeurs pour expliquer pourquoi on les met.